infer/src/absint/ProcCfg.ml

```diff
@@ -201,6 +201,98 @@ module Normal = struct
   let wto = Procdesc.get_wto
 end
 
+(** Forward .NET CFG with exceptional control-flow *)
+module DOTNETExceptional = struct
+  module Node = DefaultNode
+
+  type id_node_map = Node.t list Procdesc.IdMap.t
+
+  type t = Procdesc.t * id_node_map
+
+  type instrs_dir = Instrs.not_reversed
+
+  let instrs = Procdesc.Node.get_instrs
+
+  let fold_normal_succs _ n ~init ~f = n |> Procdesc.Node.get_succs |> List.fold ~init ~f
+
+  let fold_normal_preds _ n ~init ~f = n |> Procdesc.Node.get_preds |> List.fold ~init ~f
+  
+  let fold_exceptional_succs _ n ~init ~f = n |> Procdesc.Node.get_exn |> List.fold ~init ~f
+
+  let fold_exceptional_preds (_, exn_pred_map) n ~init ~f =
+    match Procdesc.IdMap.find (Procdesc.Node.get_id n) exn_pred_map with
+    | exn_preds ->
+        List.fold exn_preds ~init ~f
+    | exception Caml.Not_found ->
+        init
+
+  let fold_normal_or_exn_succs fold_normal_alpha fold_exceptional t n ~init ~f =
+    let choose_normal_or_exn_succs node = 
+      let instrs = Procdesc.Node.get_instrs node in
+      let instr_count = (Instrs.count instrs) - 3 in
+      let get_last_instr () =
+         if instr_count >= 0 then Instrs.nth_exn instrs instr_count else Instrs.last instrs |> Option.value ~default:Sil.skip_instr
+      in
+      let last_instr = get_last_instr () in
+      match last_instr with 
+      | Sil.Store {e1= Exp.Lvar pvar; e2= Exp.Exn _} when Pvar.is_return pvar ->
```

Comment:
This corresponds to calls to `throw e` that are translated by the (Java and C#?) frontends as just `ret = Exn e`, right? Worth adding a comment about what this is meant to capture and also that this is the only difference with `Exceptional` (unless I missed other changes) as there is an unfortunate amount of code duplication.

Reply:
Done.

Reply:
Done

Reply:
Yeah, maybe we can create an Exceptional interface and inherit? In this way we can probably avoid most of the code duplication. Please let me know what you think.